src/Compilers/CSharp/Portable/Symbols/OverriddenOrHiddenMembersHelpers.cs

```diff
@@ -920,43 +905,85 @@ private static int CustomModifierCount(Symbol member)
         /// <summary>
         /// Determine if this method requires a methodimpl table entry to inform the runtime of the override relationship.
         /// </summary>
-        internal static bool RequiresExplicitOverride(this MethodSymbol method)
+        /// <param name="warnAmbiguous">True if we should produce an ambiguity warning per https://github.com/dotnet/roslyn/issues/45453 .</param>
+        internal static bool RequiresExplicitOverride(this MethodSymbol method, out bool warnAmbiguous)
+        {
+            warnAmbiguous = false;
+            if (!method.IsOverride)
+                return false;
+
+            MethodSymbol csharpOverriddenMethod = method.OverriddenMethod;
+            if (csharpOverriddenMethod is null)
+                return false;
+
+            MethodSymbol runtimeOverriddenMethod = method.GetFirstRuntimeOverriddenMethodIgnoringNewSlot(out bool wasAmbiguous);
+            if (csharpOverriddenMethod == runtimeOverriddenMethod && !wasAmbiguous)
+                return false;
+
+            // See https://github.com/dotnet/roslyn/issues/45453. No need to warn when the runtime
+            // supports covariant returns because the methodimpl is unambiguously understood by the runtime.
+            if (method.ContainingAssembly.RuntimeSupportsCovariantReturnsOfClasses)
+                return true;
+
+            // If the method was declared as a covariant return, there will be a compile-time error since the runtime
+            // does not support covariant returns. In this case we do not warn about runtime ambiguity and pretend that
+            // we can use a methodimpl (even though it is of a form not supported by the runtime and would result in a
+            // loader error) so that the symbol APIs produce the most useful result.
+            if (!method.ReturnType.Equals(csharpOverriddenMethod.ReturnType, TypeCompareKind.AllIgnoreOptions))
+                return true;
+
+            // Due to https://github.com/dotnet/runtime/issues/38119 the methodimpl would
+            // appear to the runtime to be ambiguous in some cases.
+            bool methodimplWouldBeAmbiguous = csharpOverriddenMethod.MethodHasRuntimeCollisionInSubstututedForm();
+            if (!methodimplWouldBeAmbiguous)
+                return true;
+
+            // We produce the warning when a methodimpl would be required but would be ambiguous to the runtime
+            warnAmbiguous = true;
+
+            bool overridenMethodContainedInSameTypeAsRuntimeOverriddenMethod =
+                csharpOverriddenMethod.ContainingType.Equals(runtimeOverriddenMethod.ContainingType, TypeCompareKind.CLRSignatureCompareOptions);
```

Comment:
>csharpOverriddenMethod.ContainingType.Equals(runtimeOverriddenMethod.ContainingType, TypeCompareKind.CLRSignatureCompareOptions); [](start = 16, length = 129)

Would it make sense to perform this check before ```MethodHasRuntimeCollisionInSubstututedForm``` check? In other words, is it possible to have a RuntimeCollisionInSubstututedForm when ```!overridenMethodContainedInSameTypeAsRuntimeOverriddenMethod```?

Reply:
> Would it make sense to perform this check before MethodHasRuntimeCollisionInSubstututedForm check? In other words, is it possible to have a RuntimeCollisionInSubstututedForm when !overridenMethodContainedInSameTypeAsRuntimeOverriddenMethod?

After reviewing implementation of MethodHasRuntimeCollisionInSubstututedForm, I see that these checks are orthogonal. However, the order could be significant in order to preserve legacy compiler behavior, especially for scenarios when we did produce a MethodImpl, but now wouldn't produce it due to ```methodimplWouldBeAmbiguous``` check. It might be worth following up on that separately.  

---
In reply to: [448079933](https://github.com/dotnet/roslyn/pull/44025#discussion_r448079933) [](ancestors = 448079933)

Reply:
To clarify, I am referring to the logic we used to have here before the Covariant Returns feature work. I.e.
```
                MethodSymbol runtimeOverriddenMethod = method.GetFirstRuntimeOverriddenMethodIgnoringNewSlot(ignoreInterfaceImplementationChanges: true);
                return csharpOverriddenMethod != runtimeOverriddenMethod &&
                    method.IsAccessor() != runtimeOverriddenMethod.IsAccessor();
```

I think we should consider simply duplicating/preserving the logic and returning true when it would produce true. 

---
In reply to: [448082197](https://github.com/dotnet/roslyn/pull/44025#discussion_r448082197) [](ancestors = 448082197,448079933)

Reply:
The order of operations here simplifies the computation of when to produce a warning (`warnAmbiguous`).

> scenarios when we did produce a MethodImpl, but now wouldn't produce it due to methodimplWouldBeAmbiguous check

The previous compilers (e.g. Roslyn master) only produced a methodimpl if the first runtime match is an accessor (e.g. `Prop.get`) and the overridden method (e.g. `get_Prop()`) is not, or vice versa. There are three ways that can happen
1. if they are contained in different types. This code handles that with the final test and produces a methodimpl.
2. if they have the same return type after generic substitution and appear in the same type. That is a compile-time error CS0082.
3. similar to case (2) but with conflicting members declared in metadata, possibly using indexed properties. 

Only case (3) could make a difference in generated code, and I believe it is not a realistic scenario. If you think it is important, however, I could preserve the test from the old code for compat purposes.

---
In reply to: [448082197](https://github.com/dotnet/roslyn/pull/44025#discussion_r448082197) [](ancestors = 448082197,448079933)

Reply:
> Only case (3) could make a difference in generated code, and I believe it is not a realistic scenario. If you think it is important, however, I could preserve the test from the old code for compat purposes.

I would be more comfortable if we preserve the original logic for compat purposes. The area is very complicated and, I think it is better be safe than sorry, in case we forgot to think about some scenario. 

---
In reply to: [448473120](https://github.com/dotnet/roslyn/pull/44025#discussion_r448473120) [](ancestors = 448473120,448082197,448079933)

Reply:
I have added the old condition and tests that demonstrate that it is used. As explained, all of those test scenarios are compile-time errors.

---
In reply to: [448493699](https://github.com/dotnet/roslyn/pull/44025#discussion_r448493699) [](ancestors = 448493699,448473120,448082197,448079933)